beacon-chain/sync/initial-sync/service.go

```diff
@@ -445,57 +386,50 @@ func (s *Service) fetchOriginColumns(pids []peer.ID) error {
 	if err != nil {
 		return err
 	}
-	req, err := s.missingColumnRequest(rob, s.cfg.BlobStorage)
+	missingColumns, err := sync.FindMissingDataColumns(
+		r,
+		rob,
+		s.cfg.P2P.NodeID(),
+		peerdas.CustodyGroupCount(),
+		s.cfg.BlobStorage,
+	)
 	if err != nil {
 		return err
 	}
-	if len(req) == 0 {
+	if len(missingColumns) == 0 {
 		log.WithField("root", fmt.Sprintf("%#x", r)).Debug("All columns for checkpoint block are present")
 		return nil
 	}
-	shufflePeers(pids)
-	pids, err = s.cfg.P2P.AdmissibleCustodyGroupsPeers(pids)
+	sidecars, err := sync.RequestDataColumnSidecars(
+		s.ctx,
+		missingColumns,
+		rob,
+		r,
+		pids,
+		s.clock,
+		s.cfg.P2P,
+		s.ctxMap,
+		s.newDataColumnsVerifier,
+	)
 	if err != nil {
 		return err
 	}
-	for i := range pids {
-		sidecars, err := sync.SendDataColumnSidecarsByRootRequest(s.ctx, s.clock, s.cfg.P2P, pids[i], s.ctxMap, &req)
-		if err != nil {
-			continue
-		}
-		if len(sidecars) != len(req) {
-			continue
-		}
-
-		wrappedBlockDataColumns := make([]verify.WrappedBlockDataColumn, 0, len(sidecars))
-		for _, sidecar := range sidecars {
-			wrappedBlockDataColumn := verify.WrappedBlockDataColumn{
-				ROBlock:      blk.Block(),
-				RODataColumn: sidecar,
-			}
-
-			wrappedBlockDataColumns = append(wrappedBlockDataColumns, wrappedBlockDataColumn)
-		}
-
-		if err := verify.DataColumnsAlignWithBlock(wrappedBlockDataColumns, s.newDataColumnsVerifier); err != nil {
-			return errors.Wrap(err, "data columns align with block")
-		}
 
-		avs := das.NewLazilyPersistentStoreColumn(s.cfg.BlobStorage)
-		current := s.clock.CurrentSlot()
-		if err := avs.PersistColumns(current, sidecars...); err != nil {
-			return err
-		}
+	// FIXME: It's not clear that the caching layer is doing anything here or in
+	// fetchOriginBlobs, which is presumably where this logic was derived from.
+	avs := das.NewLazilyPersistentStoreColumn(s.cfg.BlobStorage)
+	current := s.clock.CurrentSlot()
+	if err := avs.PersistColumns(current, sidecars...); err != nil {
+		return err
+	}
 
-		nodeID := s.cfg.P2P.NodeID()
-		if err := avs.IsDataAvailable(s.ctx, nodeID, current, rob); err != nil {
-			log.WithField("root", fmt.Sprintf("%#x", r)).WithField("peerID", pids[i]).Warn("Columns from peer for origin block were unusable")
-			continue
-		}
-		log.WithField("nColumns", len(sidecars)).WithField("root", fmt.Sprintf("%#x", r)).Info("Successfully downloaded blobs for checkpoint sync block")
-		return nil
+	nodeID := s.cfg.P2P.NodeID()
+	if err := avs.IsDataAvailable(s.ctx, nodeID, current, rob); err != nil {
+		return fmt.Errorf("couldn't assemble the required columns from peers for checkpoint sync block %#x", r)
 	}
-	return fmt.Errorf("no connected peer able to provide columns for checkpoint sync block %#x", r)
+
+	log.WithField("nColumns", len(sidecars)).WithField("root", fmt.Sprintf("%#x", r)).Info("Successfully downloaded blobs for checkpoint sync block")
```

Comment:
The message should state data columns, not blobs.

Reply:
What happens if some of the peers in `RequestDataColumnSidecars` did not return the data columns?
Unless I'm mistaken, this function still exits on success.

==> We should check the all requested data columns are here, and retry until it's OK. (With a maximum trials count.)